aio: remove gulp tasks

petebacondarwin opened the pull request:
This is built upon #14868. Only the last commit is specific to this PR.

Closes #14921

petebacondarwin committed:
refactor(aio): reorganize search service for easier testing

petebacondarwin committed:
test(aio): add placeholder test descriptions

petebacondarwin committed:
test(aio): reimplemented all the commented-out unit tests

petebacondarwin committed:
build(aio): upgrade to angular 4.0.0-rc.2

IgorMinar commented on aio/src/app/shared/link.directive.spec.ts:
```diff
@@ -1,4 +1,7 @@
 import { LinkDirective } from './link.directive';
 
-describe('LinkDirective', () => {
+xdescribe('LinkDirective', () => {
```
don't use `xdescribe` simply remove the second argument from all nested `it`s and you'll get the same effect but the its will be marked as pending rather than disabled.

petebacondarwin replied:
There is a bug in zone.js. See my other comment

IgorMinar commented on aio/src/app/navigation/navigation.service.spec.ts:
```diff
@@ -1,14 +1,34 @@
+  });
+
+  xit('should fetch the navigation views', () => {});
+  xit('should compute the navigation map', () => {});
```
just remove the second arg

petebacondarwin replied:
@IgorMinar  It doesn't work because there is a bug in zone.js:

```
  /**
   * Gets a function wrapping the body of a Jasmine `it/beforeEach/afterEach` block to
   * execute in a ProxyZone zone.
   * This will run in `testProxyZone`. The `testProxyZone` will be reset by the `ZoneQueueRunner`
   */
  function wrapTestInZone(testBody: Function): Function {
    // The `done` callback is only passed through if the function expects at least one argument.
    // Note we have to make a function with correct number of arguments, otherwise jasmine will
    // think that all functions are sync or async.
    return (testBody.length == 0) ? function() {
      return testProxyZone.run(testBody, this);
    } : function(done) {
      return testProxyZone.run(testBody, this, [done]);
    };
  }
```

Note that it doesn't check to see if the function body is defined before reading its length.

petebacondarwin replied:
I have a PR for it: https://github.com/angular/zone.js/pull/659
But the zone.js project has problems of its own :-(

IgorMinar commented on aio/package.json:
```diff
@@ -7,17 +7,19 @@
-    "lint": "yarn run check-env && ng lint",
+    "ng": "yarn check-env && ng",
+    "start": "yarn check-env && ng serve",
+    "build": "yarn check-env && yarn docs && ng build -prod -aot -sm",
```
- `-aot` the default

gkalpak replied:
Does it work with single dashes? I thought you needed double dash for non-shortcut options.

gkalpak replied:
Should we have separate yarn scripts for dev build vs prod build? Or do we always want the prod build?

petebacondarwin replied:
Why would be "build" a dev version?
If I am doing development work then I use "ng serve"

gkalpak replied:
Point taken :smiley:

IgorMinar commented on aio/package.json:
```diff
@@ -7,17 +7,19 @@
-    "pre~~deploy": "yarn run check-env && ng build --prod",
+    "e2e": "yarn check-env && ng e2e --no-webdriver-update",
+    "deploy-staging": "firebase use staging --token \"$FIREBASE_TOKEN\" && yarn ~~deploy",
+    "pre~~deploy": "yarn check-env && ng build --prod",
```
is it --prod or -prod?

petebacondarwin replied:
From the Angular CLI help usage:

```
--target (String) (Default: development) Defines the build target.
    aliases: -t <value>, -dev (--target=development), -prod (--target=production), --target <value>, --target <value>
```

So it appears to be `-prod`. I'll fix this

gkalpak replied:
The search results indicate otherwise: [--prod][1] vs [-prod][2]
Could it be a typo in the help message?

[1]: https://github.com/angular/angular-cli/search?utf8=%E2%9C%93&q=--prod&type=Code
[2]: https://github.com/angular/angular-cli/search?utf8=%E2%9C%93&q=-prod&type=Code

IgorMinar reviewed (changes requested):
the CI is failing

gkalpak commented on aio/check-env.js:
```diff
@@ -0,0 +1,6 @@
+  requiredNodeVersion: engines.node,
+  requiredNpmVersion: engines.npm,
+  requiredYarnVersion: engines.yarn
+});
```
Why do we need a separate `check-env.js`?
Is it so that we don't rely on stuff outside the `aio/` directory?

gkalpak replied:
BTW, I would prefer if such scripts were not on the root-level. We already have a `scripts/` directory :wink:

petebacondarwin replied:
we don't have a `scripts` folder in `aio/`.
I am cool with moving this to the top level folder.

gkalpak replied:
Right. We don't have right now. But we will have one when #14330 lands :wink:

gkalpak commented on aio/package.json:
```diff
@@ -7,17 +7,19 @@
-    "deploy-staging": "firebase use staging --token \"$FIREBASE_TOKEN\" && yarn run ~~deploy",
-    "pre~~deploy": "yarn run check-env && ng build --prod",
+    "e2e": "yarn check-env && ng e2e --no-webdriver-update",
+    "deploy-staging": "firebase use staging --token \"$FIREBASE_TOKEN\" && yarn ~~deploy",
```
Wow! We don't need `run` anymore with yarn? 🕺

petebacondarwin replied:
Yes that is something that `yarn` really got right!

petebacondarwin committed:
build: allow check-environment.js to be run directly

petebacondarwin committed:
build(aio): move gulp tasks to package.json

Generate the docs with `yarn docs`.
Test the doc generation code with `yarn docs-test`

The docs are automatically built as part of the `yarn build` task,
so there is no need to rebuild them in the test_aio.sh file

petebacondarwin commented:
Addressed all comments, rebased on master, aio tasks in the Travis build now pass.

IgorMinar commented:
@petebacondarwin I rebased your PR and pushed to your branch. I'm about to merge this in if CI passes, so please do not make more changes in this branch
